### ISSUE TITLE
Support needed for bare word file handles and open()

### ISSUE THREAD

**Issue by drmuey** [cpanel/Test-MockFile#34]
perl v5.28.0, Test-MockFile v0.016 (that is what cpanel 80 is set to ATM)

## Modulino w/ Config::Tiny

```perl
#!/usr/local/cpanel/3rdparty/bin/perl
package modulino;

use strict;
use warnings;

use Data::Dumper;
require Config::Tiny;
 
sub main {
    my $cfg = load("$ENV{HOME}/.modulino.conf");
    print Dumper($cfg);
    return;
}
 
sub load { return Config::Tiny->read(shift) }
 
main() unless caller;
 
1;
```

## Test

```perl
use strict;
use warnings;
 
use Test::More tests => 1;
use FindBin;
 
use Test::MockFile;
require "$FindBin::Bin/modulino";
 
my $file        = "$ENV{HOME}/.modulino.conf";
my $mocked_conf = Test::MockFile->file( $file, "[derp]\nfoo=2\nbar=7\n" );
my $exp         = { derp => { foo => 2, bar => 7 } };
 
# diag explain modulino::load($file);
is_deeply modulino::load($file), $exp, "can load ini file";
```

## Result

```
# prove -vw modulino.t
modulino.t .. 
1..1
Modification of a read-only value attempted at /usr/local/cpanel/3rdparty/perl/528/lib/perl5/cpanel_lib/Test/MockFile.pm line 1070.
# Looks like your test exited with 255 before it could output anything.
```
Change the modulino to use open() (i.e. not Config::Tiny) adjust test accordingly and it passes:

## Modulino w/out Config::Tiny

```perl
#!/usr/local/cpanel/3rdparty/bin/perl
package modulino;

use strict;
use warnings;

use Data::Dumper;

sub main {
    my $cfg = load("$ENV{HOME}/.modulino.conf");
    print Dumper($cfg);
    return;
}

sub load {
    my ($file) = @_;
    if ( open my $fh, "<", $file ) {
        return <$fh>;
    }
    return;
}

main() unless caller;

1;
```

## Test

```perl
use strict;
use warnings;

use Test::More tests => 1;
use FindBin;

use Test::MockFile;
require "$FindBin::Bin/modulino";

my $file        = "$ENV{HOME}/.modulino.conf";
my $cont        = "oh hai $$\n";
my $mocked_conf = Test::MockFile->file( $file, $cont );

# diag explain modulino::load($file);
is_deeply modulino::load($file), $cont, "can load file";
```

## Result

```
# prove -w modulino.t
modulino.t .. ok   
All tests successful.
Files=1, Tests=1,  0 wallclock secs ( 0.04 usr  0.01 sys +  0.10 cusr  0.01 csys =  0.16 CPU)
Result: PASS
```

**Comment by atoomic** [comment]
side note, `strict` and `warnings` would need to be enabled after declaring the `package`

**Comment by atoomic** [comment]
the problem comes from Config::Tiny::read which is using a bare file handle for the open, which is currently not supported

patching it with something like this solves the issue
```perl
 sub read
 {
     my($class)           = ref $_[0] ? ref shift : shift;
     my($file, $encoding) = @_;

     return $class -> _error('No file name provided') if (! defined $file || ($file eq '') );

     # Slurp in the file.

     $encoding = $encoding ? "<:$encoding" : '<';
     local $/  = undef;

     my $fh;
     open( $fh, $encoding, $file ) or return $class -> _error( "Failed to open file '$file' for reading: $!" );
     my $contents = readline $fh;
     close( $fh );

     return $class -> _error("Reading from '$file' returned undef") if (! defined $contents);

     return $class -> read_string( $contents );

 } # End of read.
```

**Comment by drmuey** [comment]
ah gotcha, thanks @atoomic much appreciated!

Until we can implement bare handle support (or if its not possible) perhaps a caveat section could document that.

That way folks using `Test::MockFile` w/ module’s they don’t have any control over won’t be lost.

**Comment by toddr** [comment]
Actually most of the code supports bare word file handles. We just need to do open

**Comment by drmuey** [comment]
> Actually most of the code supports bare word file handles. We just need to do open

ok thanks, I'll update the pull request then

**Comment by drmuey** [comment]
#35 PR updated to limit it to `open()`

**Comment by toddr** [comment]
While we're at it, let's submit a PR since using global file handles isn't generally a good idea.

https://github.com/ronsavage/Config-Tiny/pull/5

**Comment by xsawyerx** [comment]
No need to document this as a limitation, I've made it work in #66.

**Comment by toddr** [comment]
Fixed by #34
